azurelinuxagent/ga/remoteaccess.py

```diff
@@ -102,60 +102,78 @@ def run(self):
                       message=msg)
 
     def handle_remote_access(self):
+        # Get JIT user accounts.
+        all_users = self.os_util.get_users()
+        existing_jit_users = [u[0] for u in all_users if self.validate_jit_user(u[4])]
+        self.err_message = ""
         if self.remote_access is not None:
-            # Get JIT user accounts.
-            all_users = self.os_util.get_users()
-            jit_users = set()
-            for usr in all_users:
-                if self.validate_jit_user(usr[4]):
-                    jit_users.add(usr[0])
             for acc in self.remote_access.user_list.users:
-                raw_expiration = acc.expiration
-                account_expiration = datetime.strptime(raw_expiration, REMOTE_USR_EXPIRATION_FORMAT)
-                now = datetime.utcnow()
-                if acc.name not in jit_users and now < account_expiration:
-                    self.add_user(acc.name, acc.encrypted_password, account_expiration)
+                try:
+                    raw_expiration = acc.expiration
+                    account_expiration = datetime.strptime(raw_expiration, REMOTE_USR_EXPIRATION_FORMAT)
+                    now = datetime.utcnow()
+                    if acc.name not in existing_jit_users and now < account_expiration:
+                        self.add_user(acc.name, acc.encrypted_password, account_expiration)
+                    elif acc.name in existing_jit_users and now > account_expiration:
+                        # user account expired, delete it.
+                        logger.info("user {0} expired from remote_access".format(acc.name))
+                        self.remove_user(acc.name)
+                except RemoteAccessError as rae:
+                    self.err_message = self.err_message + "Error processing user {0}. Exception: {1}"\
+                        .format(acc.name, ustr(rae))
+            for user in existing_jit_users:
+                try:
+                    if user not in [u.name for u in self.remote_access.user_list.users]:
```

Comment:
Can you hoist the construction of this list (`[u.name for u in self.remote_access.user_list.users]`) outside the loop? If `self.remove_user(user)` alters `self.remote_access.user_list.users`, then recompute the object when you remove a user.